File: python/db/query_geo_crosswalk.py
#!/usr/bin/env python3
"""
Query geographic crosswalk tables from DuckDB and export to Feather format.

This script is called from R to safely access DuckDB without segmentation faults.
The hybrid Python→Feather→R approach ensures 100% reliability.

Usage:
    python python/db/query_geo_crosswalk.py --table geo_zip_to_puma --output temp.feather
"""

import sys
import argparse
import duckdb
import pandas as pd
from pathlib import Path


def query_crosswalk(table_name: str, output_path: str, db_path: str = None) -> bool:
    """
    Query a geographic crosswalk table and write to Feather format.

    Args:
        table_name: Name of crosswalk table to query
        output_path: Path for output Feather file
        db_path: Path to DuckDB database (default: data/duckdb/kidsights_local.duckdb)

    Returns:
        True if successful, False otherwise
    """
    if db_path is None:
        db_path = "data/duckdb/kidsights_local.duckdb"

    try:
        # Connect to database
        conn = duckdb.connect(db_path, read_only=True)

        # Query table
        df = conn.execute(f"SELECT * FROM {table_name}").fetchdf()

        conn.close()

        if df.empty:
            print(f"WARNING: Table {table_name} is empty", file=sys.stderr)
            return False

        # Write to Feather
        df.to_feather(output_path)

        # Success message to stdout
        print(f"SUCCESS: Exported {len(df)} rows from {table_name} to {output_path}")

        return True

    except Exception as e:
        print(f"ERROR: {e}", file=sys.stderr)
        return False


def main():
    """Main entry point."""
    parser = argparse.ArgumentParser(
        description='Query geographic crosswalk from DuckDB'
    )
    parser.add_argument(
        '--table',
        required=True,
        help='Table name to query'
    )
    parser.add_argument(
        '--output',
        required=True,
        help='Output Feather file path'
    )
    parser.add_argument(
        '--db',
        default='data/duckdb/kidsights_local.duckdb',
        help='Database path (default: data/duckdb/kidsights_local.duckdb)'
    )

    args = parser.parse_args()

    success = query_crosswalk(
        table_name=args.table,
        output_path=args.output,
        db_path=args.db
    )

    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: R/utils/query_geo_crosswalk.R
#' Query Geographic Crosswalk from DuckDB
#'
#' Safely queries geographic crosswalk reference tables from DuckDB using
#' the hybrid Python→Feather→R approach to avoid segmentation faults.
#'
#' @param table_name Name of the crosswalk table to query (e.g., "geo_zip_to_puma")
#' @param temp_dir Directory for temporary Feather files (default: tempdir())
#' @param python_path Path to Python executable (default: python)
#'
#' @return Data frame with crosswalk data, or NULL on error
#'
#' @examples
#' \dontrun{
#' puma_data <- query_geo_crosswalk("geo_zip_to_puma")
#' county_data <- query_geo_crosswalk("geo_zip_to_county")
#' }
#'
#' @export
query_geo_crosswalk <- function(
  table_name,
  temp_dir = NULL,
  python_path = "python"
) {

  # Use default temp dir if not specified
  if (is.null(temp_dir)) {
    temp_dir <- tempdir()
  }

  # Create unique temp file path
  temp_file <- file.path(temp_dir, paste0(table_name, "_", Sys.getpid(), ".feather"))

  # Clean up temp file on exit
  on.exit({
    if (file.exists(temp_file)) {
      unlink(temp_file)
    }
  }, add = TRUE)

  # Build Python command
  python_script <- "python/db/query_geo_crosswalk.py"

  # Check if script exists
  if (!file.exists(python_script)) {
    stop(paste("Python script not found:", python_script))
  }

  cmd <- sprintf(
    '%s %s --table "%s" --output "%s"',
    python_path,
    python_script,
    table_name,
    temp_file
  )

  # Execute Python script
  result <- system(cmd, intern = TRUE, ignore.stderr = FALSE)

  # Check if temp file was created
  if (!file.exists(temp_file)) {
    stop(paste("Failed to query", table_name, "from database"))
  }

  # Read Feather file
  crosswalk_data <- arrow::read_feather(temp_file)

  # Validate result
  if (nrow(crosswalk_data) == 0) {
    warning(paste("Table", table_name, "is empty"))
    return(NULL)
  }

  return(crosswalk_data)
}


#' Get Available Geographic Crosswalk Tables
#'
#' Returns a character vector of available geographic crosswalk table names.
#'
#' @return Character vector of table names
#'
#' @export
get_geo_crosswalk_tables <- function() {
  c(
    "geo_zip_to_puma",
    "geo_zip_to_county",
    "geo_zip_to_tract",
    "geo_zip_to_cbsa",
    "geo_zip_to_urban_rural",
    "geo_zip_to_school_dist",
    "geo_zip_to_state_leg_lower",
    "geo_zip_to_state_leg_upper",
    "geo_zip_to_congress",
    "geo_zip_to_native_lands"
  )
}
